staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go

```diff
@@ -142,13 +180,40 @@ func (t *envelopeTransformer) TransformToStorage(data []byte, context value.Cont
 	// Append the length of the encrypted DEK as the first 2 bytes.
 	b := cryptobyte.NewBuilder(nil)
 	b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-		b.AddBytes([]byte(encKey))
+		b.AddBytes(dekCiphertext)
 	})
 	b.AddBytes(result)
 
 	return b.Bytes()
 }
 
+func (t *envelopeTransformer) rotate() ([]byte, value.Transformer, error) {
+	var (
+		err error
+	)
+
+	t.mux.RLock()
+	if !t.currentDEK.shouldRotate() {
+		t.mux.RUnlock()
+		return t.currentDEK.ciphertext, t.currentDEK.transformer, nil
+	}
+	t.mux.RUnlock()
+
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	t.currentDEK, err = newDEK(t.envelopeService, t.dekCacheTTL, t.baseTransformerFunc)
```

Comment:
check `shouldRotate` here again, before calling `newDEK`.
another goroutine could've rotated in between your read lock and write lock

Reply:
Done.